base/compiler/typeinfer.jl

```diff
@@ -659,6 +659,19 @@ function typeinf_ext_toplevel(interp::AbstractInterpreter, linfo::MethodInstance
 end
 
 
+const docstring_return_type =
+"""
+    return_type(f, ArgType)
+
+Estimate the return type of `f` given arguments of type `ArgType`. In other words this gives an upper bound for `typeof(f((args::ArgType)...))`.
+
+!!! warning
+
+    This function is not part of the stable API. It is an implementation detail of the compiler,
+    that can and will change any time. For instance it might give sharper bounds as
+    inference improves or looser bounds to improve latency.
+    Using `return_type` is very likely to lead to undefined behaviour.
```

Comment:
Are there more reasons to avoid `return_type`? I vaguely remember, that it hurts inference, is that true?

Reply:
"Undefined behavior" has a very specific meaning in programming languages. I would say "unpredictable behavior" instead.

This pretty much covers the reasons not to use it, but you could also add that it makes programs less robust since this function is likely to do what you want only for certain types and not all types.